nano/core_test/confirmation_height.cpp

```diff
@@ -1,2187 +0,0 @@
-
-		nano::block_hash latest (node->latest (nano::dev::genesis_key.pub));
-
-		nano::keypair key1;
-		nano::block_builder builder;
-		auto send = builder
-					.send ()
-					.previous (latest)
-					.destination (key1.pub)
-					.balance (nano::dev::constants.genesis_amount - nano::Gxrb_ratio)
-					.sign (nano::dev::genesis_key.prv, nano::dev::genesis_key.pub)
-					.work (*system.work.generate (latest))
-					.build ();
-		auto send1 = builder
-					 .send ()
-					 .previous (send->hash ())
-					 .destination (key1.pub)
-					 .balance (nano::dev::constants.genesis_amount - nano::Gxrb_ratio * 2)
-					 .sign (nano::dev::genesis_key.prv, nano::dev::genesis_key.pub)
-					 .work (*system.work.generate (send->hash ()))
-					 .build ();
-		auto send2 = builder
-					 .send ()
-					 .previous (send1->hash ())
-					 .destination (key1.pub)
-					 .balance (nano::dev::constants.genesis_amount - nano::Gxrb_ratio * 3)
-					 .sign (nano::dev::genesis_key.prv, nano::dev::genesis_key.pub)
-					 .work (*system.work.generate (send1->hash ()))
-					 .build ();
-		{
-			auto transaction = node->store.tx_begin_write ();
-			ASSERT_EQ (nano::block_status::progress, node->ledger.process (transaction, send));
-			ASSERT_EQ (nano::block_status::progress, node->ledger.process (transaction, send1));
-			ASSERT_EQ (nano::block_status::progress, node->ledger.process (transaction, send2));
-		}
-
-		add_callback_stats (*node);
-
-		// This election should be confirmed as active_conf_height
-		ASSERT_TRUE (nano::test::start_election (system, *node, send1->hash ()));
-		// Start an election and confirm it
-		auto election = nano::test::start_election (system, *node, send2->hash ());
-		ASSERT_NE (nullptr, election);
-		election->force_confirm ();
-
-		ASSERT_TIMELY_EQ (5s, node->stats.count (nano::stat::type::http_callback, nano::stat::detail::http_callback, nano::stat::dir::out), 3);
-
-		ASSERT_EQ (1, node->stats.count (nano::stat::type::confirmation_observer, nano::stat::detail::active_quorum, nano::stat::dir::out));
-		ASSERT_EQ (1, node->stats.count (nano::stat::type::confirmation_observer, nano::stat::detail::active_conf_height, nano::stat::dir::out));
-		ASSERT_EQ (1, node->stats.count (nano::stat::type::confirmation_observer, nano::stat::detail::inactive_conf_height, nano::stat::dir::out));
-		ASSERT_EQ (3, node->stats.count (nano::stat::type::confirmation_height, nano::stat::detail::blocks_confirmed, nano::stat::dir::in));
-		ASSERT_EQ (3, node->stats.count (nano::stat::type::confirmation_height, get_stats_detail (mode_a), nano::stat::dir::in));
-		ASSERT_EQ (4, node->ledger.cache.cemented_count);
-
-		ASSERT_EQ (0, node->active.election_winner_details_size ());
-	};
-
-	test_mode (nano::confirmation_height_mode::bounded);
-	test_mode (nano::confirmation_height_mode::unbounded);
-}
-
-// This test checks that a receive block with uncemented blocks below cements them too.
-TEST (confirmation_height, cemented_gap_below_receive)
```

Comment:
I put these test deletions into a single commit to single out their removal. https://github.com/nanocurrency/nano-node/pull/4519/commits/abc127ad5999530e738da2bcbe72a201360ec055

These tests seemed to be testing behavior unrelated to confirming: gaps in source/previous or bootstrapping/live block sources. I'm not fully sure the original intent whether it was pedantic or it might have to do with multiple db transactions being used per traversal.

The new implementation is strongly intolerant to missing blocks if they are requested to be confirmed and will release_assert.

Reply:
Okay, this makes total sense. I was confused by those tests and didn't spend too much time analyzing them, thanks for the explanation. I agree that crashing loudly with release asserts is the way to go in such cases, if chains with gaps are being cemented then something must have gone terribly wrong.